src/main/java/fi/aalto/cs/apluscourses/intellij/actions/FilterOptionsActionGroup.java

```diff
@@ -0,0 +1,40 @@
+package fi.aalto.cs.apluscourses.intellij.actions;
+
+import com.intellij.openapi.actionSystem.ActionGroup;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import fi.aalto.cs.apluscourses.intellij.services.MainViewModelProvider;
+import fi.aalto.cs.apluscourses.intellij.services.PluginSettings;
+import fi.aalto.cs.apluscourses.presentation.MainViewModel;
+import fi.aalto.cs.apluscourses.presentation.base.BaseTreeViewModel;
+import fi.aalto.cs.apluscourses.utils.Streamable;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class FilterOptionsActionGroup extends ActionGroup {
+
+  private final MainViewModelProvider mainViewModelProvider;
+
+  public FilterOptionsActionGroup() {
+    this(PluginSettings.getInstance());
+  }
+
+  public FilterOptionsActionGroup(MainViewModelProvider mainViewModelProvider) {
+    this.mainViewModelProvider = mainViewModelProvider;
+  }
+
+  // Caution: Array covariance in the return type.
```

Comment:
What's the array covariance here? Doesn't the method create an array `FilterOptionAction[]`?

Reply:
Yes it does. What I meant was that `FilterOptionsActionGroup` overrides a method returning `AnAction[]` with a method returning `FilterOptionAction[]` (and why is that? for convenience in testing). In Java, that is allowed, and it shouldn't cause any problems here because the caller is not expected to modify the array.